File: huffman/src/c++/classes/huffman.cpp
#include "huffman.hpp"

Node huffman(vector<Node>& nodes) {
  MinPriorityQueue queue(nodes);

  cout << "Queue is empty = " << queue.is_empty() << endl << endl;
  cout << "Queue" << endl;
  queue.print_queue();
  cout << endl;

  while (!queue.is_empty()) {
    Node node;

    node.set_left_child(queue.extract_min());
    node.set_right_child(queue.extract_min());

    node.set_frequency(node.get_left_child()->get_frequency() + node.get_right_child()->get_frequency());

    cout << "Node: " << node.get_node() << endl << endl;

    queue.insert(node);
  }

  return *(queue.extract_min());
}

File: huffman/src/python/huffman.py
from typing import List
from node import Node
from min_priority_queue import Min_Priority_Queue

def huffman(nodes: List[Node]) -> Node:
    queue = Min_Priority_Queue(nodes)
        
    while queue.size() > 1:
        new_node = Node(None, 0, None, None, True)
        
        new_node.left_child = queue.extract_min()
        new_node.right_child = queue.extract_min()
        
        new_node.frequency = new_node.left_child.frequency + new_node.right_child.frequency
        
        queue.push(new_node)
      
    return print(f"root: {queue.peek()}")
  
nodes = [
    Node("A", 5), 
    Node("B", 10), 
    Node("C", 15), 
    Node("D", 20), 
    Node("E", 25),
    Node("F", 25)
]
huffman(nodes)